opentracing-apache-httpclient/src/main/java/io/opentracing/contrib/apache/TracingClientExec.java

```diff
@@ -0,0 +1,151 @@
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.http.HttpException;
+import org.apache.http.HttpRequest;
+import org.apache.http.client.RedirectStrategy;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpExecutionAware;
+import org.apache.http.client.methods.HttpRequestWrapper;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.conn.routing.HttpRoute;
+import org.apache.http.impl.execchain.ClientExecChain;
+
+/**
+ * Tracing is added before {@link org.apache.http.impl.execchain.ProtocolExec} which is invoked as
+ * the next to last. Note that {@link org.apache.http.impl.execchain.RedirectExec} is invoked before
+ * so this exec has to deal with redirects.
+ *
+ * @author Pavol Loffay
+ */
+public class TracingClientExec implements ClientExecChain {
+  static final String COMPONENT_NAME = "apache-httpclient";
+
+  /**
+   * Id of {@link HttpClientContext#setAttribute(String, Object)} representing span associated with
+   * the current client processing. Referenced span is local span not a span representing HTTP communication.
+   */
+  protected static final String ACTIVE_SPAN = TracingHttpClientBuilder.class.getName() + ".activeSpan";
+  /**
+   * Tracing {@link ClientExecChain} is executed after redirect exec, so on redirects it is called
+   * multiple times. This is used as an id for {@link HttpClientContext#setAttribute(String, Object)}
+   * to store number of redirects.
+   */
+  protected static final String REDIRECT_COUNT = TracingHttpClientBuilder.class.getName() + ".redirectCount";
+
+  private final RedirectStrategy redirectStrategy;
+  private final ClientExecChain requestExecutor;
+  private final boolean redirectHandlingDisabled;
+
+  private final Tracer tracer = GlobalTracer.get();
+  private final SpanManager spanManager = DefaultSpanManager.getInstance();
+  private final List<ApacheClientSpanDecorator> spanDecorators;
+
+  public TracingClientExec(ClientExecChain clientExecChain, RedirectStrategy redirectStrategy,
+      boolean redirectHandlingDisabled, List<ApacheClientSpanDecorator> spanDecorators) {
+    this.requestExecutor = clientExecChain;
+    this.redirectStrategy = redirectStrategy;
+    this.redirectHandlingDisabled = redirectHandlingDisabled;
+    this.spanDecorators = new ArrayList<>(spanDecorators);
+  }
+
+  @Override
+  public CloseableHttpResponse execute(HttpRoute route,
+      HttpRequestWrapper request,
+      HttpClientContext clientContext,
+      HttpExecutionAware execAware) throws IOException, HttpException {
+
+    Span localSpan = clientContext.getAttribute(ACTIVE_SPAN, Span.class);
+    CloseableHttpResponse response = null;
+    try {
+      if (localSpan != null) {
+        return (response = handleNetworkProcessing(localSpan, route, request, clientContext, execAware));
+      }
+
+      localSpan = handleLocalSpan(request, clientContext);
```

Comment:
Instead of code above, why not just surround this statement by `if (localSpan == null)` ?